backend/src/v2/driver/driver.go

```diff
@@ -671,7 +671,7 @@ func resolveInputs(ctx context.Context, dag *metadata.DAG, iterationIndex *int,
 				}
 			}
 			value, hasValue := inputs.GetParameterValues()[name]
-			if !hasValue {
+			if !hasValue && inputsSpec.GetParameters()[name].IsOptional == false {
```

Comment:
With this change, aren't we missing the opportunity to get default value when `IsOptional == true`?

Reply:
This is a really good point! I should change the logic here.

Reply:
Would it be more clear? WDYT?
```golang
if !hasValue {
   value = spec.GetDefaultValue()
   if inputsSpec.GetParameters()[name].IsOptional {
      value = structpb.NewStringValue("")
   }
   if value == nil {
      return fmt.Errorf("no value or default value provided for non-optional parameter %q", name)
   }
}
```

Reply:
Thank you @gkcalat for suggesting this. I'm already using a different logic handling this, following this truth table:
| optional | has value | action                                    |
|----------|-----------|-------------------------------------------------|
| +        | +         | use provided value                              |
| +        | -         | use default value, if no default, then use None |
| -        | +         | use provided value                              |
| -        | -         | use default value, if no default, then error    |
When a parameter is optional, it is still possible that there is a default value, see [example](https://github.com/kubeflow/pipelines/blob/69969475bc09ea0d48115ea48a0b0041043c1d1c/samples/test/lightweight_python_functions_v2_with_outputs.py#L72).

Reply:
> 
> optional | has value | action
> -- | -- | --
> | + | - |  use default value, if no default, then use None |

What does this `None` mean? Is it Python `None`, or Golang `nil`, or something else?

Reply:
See [note](https://github.com/kubeflow/pipelines/pull/8765#issuecomment-1439774421).